app/models/manageiq/providers/nuage/network_manager/event_catcher/stream.rb

```diff
@@ -28,18 +33,39 @@ def initialize(options = {})
   def start(&message_handler_block)
     _log.debug("#{self.class.log_prefix} Opening amqp connection using options #{@options}")
     @options[:message_handler_block] = message_handler_block if message_handler_block
-    connection.run
+    with_fallback_urls(@options[:urls]) do
+      connection.run
+    end
   end
 
   def stop
-    @handler.stop
+    @handler&.stop
   end
 
   def connection
     unless @connection
-      @handler = ManageIQ::Providers::Nuage::NetworkManager::EventCatcher::MessagingHandler.new(@options)
-      @connection = Qpid::Proton::Reactor::Container.new(@handler)
+      @handler = ManageIQ::Providers::Nuage::NetworkManager::EventCatcher::MessagingHandler.new(@options.clone)
+      @connection = Qpid::Proton::Container.new(@handler)
     end
     @connection
   end
+
+  def with_fallback_urls(urls)
+    urls.each_with_index do |url, idx|
+      endpoint_str = "ActiveMQ endpoint #{idx + 1}/#{@options[:urls].count} (#{url})"
+      _log.info("#{self.class.log_prefix} Connecting to #{endpoint_str}")
+      begin
+        @options[:url] = url
+        yield
+      rescue MiqException::MiqHostError, Errno::ECONNREFUSED => err
+        _log.info("#{self.class.log_prefix} #{endpoint_str} errored: #{err}")
+        stop
+        reset_connection
```

Comment:
What happens when the first connection fails?

Reply:
A very good question that I had to come to by hard :) It's working fine even when first connection fails, it proceeds to the fallback. I've tested with portforwarding as suggested by you some time ago. So I've set it to `localhost:5672, localhost:5672, <actual_endpoint>` and it worked even after EMS restart (that's why we rescue also `Errno::ECONNREFUSED`).

Reply:
Two more:

* What happens when all three fail?
* What happens when one of the URLs will reappear?

Reply:
I cannot give a definite answer to what happens if all URLs are invalid, since my knowledge of this part of the code is really shallow, but my guess would be that the thread terminates and tells no one about it?

If we would like to achieve the same behavior as the old Ruby binding offered, then we should wrap lines 54:63 in infinite loop (old bindings created circular list from the urls that were passed to the connect call). Adding a loop will allow us to detect if/when one or more URLs reappear.

Reply:
I would kindly ask you to test this as it's an important feature that needs to either be implemented (loops with some delays, of course) or documented in the end.

Reply:
I've double tested and behavior is as expected: when certain endpoint dies it tries to connect to next fallback. If connection succeeds, it usines that fallback. If the fallback dies at some point, next callback is tested. If we're out of callbacks, then this is reported to log (same as when any previous endpoint died) and EventCatcher worker is restarted meaning that first endpoint is tested again.

Here is the console output that I get if I forcibly prevent Nuage from being able to access all three endpoints:
```
[----] ...::Nuage::NetworkManager::EventCatcher::Runner#start_event_monitor) EMS [...] as [...] Starting Event Monitor Thread
[----] ...::Nuage::NetworkManager::EventCatcher::Stream#with_fallback_urls) MIQ(Class) Connecting to ActiveMQ endpoint 1/3 (localhost:5672)
[----] ...::Nuage::NetworkManager::EventCatcher::Stream#with_fallback_urls) MIQ(Class) ActiveMQ endpoint 1/3 (localhost:5672) errored: Connection refused - connect(2) for "localhost" port 5672
[----] ...::Nuage::NetworkManager::EventCatcher::Stream#with_fallback_urls) MIQ(Class) Connecting to ActiveMQ endpoint 2/3 (localhost:5672)
[----] ...::Nuage::NetworkManager::EventCatcher::Stream#with_fallback_urls) MIQ(Class) ActiveMQ endpoint 2/3 (localhost:5672) errored: Connection refused - connect(2) for "localhost" port 5672
[----] ...::Nuage::NetworkManager::EventCatcher::Stream#with_fallback_urls) MIQ(Class) Connecting to ActiveMQ endpoint 3/3 (localhost:5672)
[----] ...::Nuage::NetworkManager::EventCatcher::Stream#with_fallback_urls) MIQ(Class) ActiveMQ endpoint 3/3 (localhost:5672) errored: Connection refused - connect(2) for "localhost" port 5672
[----] ...::Nuage::NetworkManager::EventCatcher::Runner#start_event_monitor) EMS [...] as [...] Started Event Monitor Thread
[----] ...::Nuage::NetworkManager::EventCatcher::Runner#do_work) EMS [...] as [...] Event Monitor Thread gone. Restarting...

[----] ...::Nuage::NetworkManager::EventCatcher::Runner#start_event_monitor) EMS [vsd01.phd.plm.nuagedemo.net] as [xlab] Starting Event Monitor Thread
[----] ...::Nuage::NetworkManager::EventCatcher::Stream#with_fallback_urls) MIQ(Class) Connecting to ActiveMQ endpoint 1/3 (localhost:5672)
[----] ...::Nuage::NetworkManager::EventCatcher::Stream#with_fallback_urls) MIQ(Class) ActiveMQ endpoint 1/3 (localhost:5672) errored: Connection refused - connect(2) for "localhost" port 5672
[----] ...::Nuage::NetworkManager::EventCatcher::Stream#with_fallback_urls) MIQ(Class) Connecting to ActiveMQ endpoint 2/3 (localhost:5672)
[----] ...::Nuage::NetworkManager::EventCatcher::Stream#with_fallback_urls) MIQ(Class) ActiveMQ endpoint 2/3 (localhost:5672) errored: Connection refused - connect(2) for "localhost" port 5672
[----] ...::Nuage::NetworkManager::EventCatcher::Stream#with_fallback_urls) MIQ(Class) Connecting to ActiveMQ endpoint 3/3 (localhost:5672)
[----] ...::Nuage::NetworkManager::EventCatcher::Stream#with_fallback_urls) MIQ(Class) ActiveMQ endpoint 3/3 (localhost:5672) errored: Connection refused - connect(2) for "localhost" port 5672
``` 

Then when I open tunnel again, effectively allowing EventCatcher to access ActiveMQ again, it connects immediately.

Reply:
This is perfect for me! Thanks @miha-plesko.